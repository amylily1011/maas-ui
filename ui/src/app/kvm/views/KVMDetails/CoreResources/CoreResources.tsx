import classNames from "classnames";

import PodMeter from "app/kvm/components/PodMeter";
import { getRanges } from "app/utils";

export type Props = {
  cores: {
    allocated: number;
    free: number;
  };
  dynamicLayout?: boolean;
  pinned?: number[];
  available?: number[];
};

const CoreResources = ({
  cores,
  dynamicLayout,
  pinned = [],
  available = [],
}: Props): JSX.Element => {
  const noPin = (pinned && pinned.length === 0) || 0;
  const pinnedRanges = getRanges(pinned).join(",");
  const noAvailable = (available && available.length === 0) || 0;
  const availableRanges = getRanges(available).join(",");
  return (
    <div
      className={classNames("core-resources", {
        "core-resources--dynamic-layout": dynamicLayout,
      })}
    >
      <h4 className="core-resources__header p-heading--small u-sv1">
        CPU cores
      </h4>
      <div className="core-resources__meter">
        <PodMeter allocated={cores.allocated} free={cores.free} segmented />
      </div>
      <div>
        {noPin && noAvailable ? (
          " "
        ) : (
          <div>
            <hr />
            <h4 className="core-resources__header p-heading--small u-sv1">
              Pinned cores
            </h4>
            <div>{noPin ? <em>None</em> : pinnedRanges}</div>
            <span className="p-text--paragraph u-text--light">
              {noAvailable
                ? "All cores are pinned."
                : `(Unpinned cores: ${availableRanges})`}
            </span>
          </div>
        )}
      </div>
    </div>
  );
};

export default CoreResources;
