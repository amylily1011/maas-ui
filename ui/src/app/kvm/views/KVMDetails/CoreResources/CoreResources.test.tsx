import { shallow } from "enzyme";

import CoreResources from "./CoreResources";

describe("CoreResources", () => {
  it("renders", () => {
    const wrapper = shallow(
      <CoreResources
        cores={{ allocated: 1, free: 2 }}
        pinned={[1, 2, 3, 4]}
        available={[0, 5, 6, 7]}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("can be made to have a dynamic layout", () => {
    const wrapper = shallow(
      <CoreResources cores={{ allocated: 1, free: 2 }} dynamicLayout />
    );

    expect(
      wrapper
        .find(".core-resources")
        .prop("className")
        ?.includes("core-resources--dynamic-layout")
    ).toBe(true);
  });
});
