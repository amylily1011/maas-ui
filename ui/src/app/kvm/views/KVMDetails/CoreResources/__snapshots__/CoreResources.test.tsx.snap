// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreResources renders 1`] = `
<div
  className="core-resources"
>
  <h4
    className="core-resources__header p-heading--small u-sv1"
  >
    CPU cores
  </h4>
  <div
    className="core-resources__meter"
  >
    <PodMeter
      allocated={1}
      free={2}
      segmented={true}
    />
  </div>
  <hr />
     <div>
       <h4 className="core-resources__header p-heading--small u-sv1">
          Pinned Cores
       </h4>
       <div>pinned={1,2,3,4}</div>
       <span className="p-text--paragraph u-text--light">
          available={0,5,6,7}       
        </span>
   </div>
</div>
`;
